src/Lucene.Net/Util/LightThreadLocal.cs

```diff
@@ -0,0 +1,144 @@
+                return ReferenceEquals(x, y);
+            }
+
+            public int GetHashCode(T obj)
+            {
+                return RuntimeHelpers.GetHashCode(obj);
+            }
+        }
+
+        public class CurrentThreadState
+        {
+            private readonly HashSet<WeakReferenceToLightThreadLocal> _parents
+                = new HashSet<WeakReferenceToLightThreadLocal>();
+
+            private class WeakReferenceToLightThreadLocal : IEquatable<WeakReferenceToLightThreadLocal>
+            {
+                private readonly WeakReference<LightThreadLocal<T>> _weak;
+                private readonly int _hashCode;
+
+                public bool TryGetTarget(out LightThreadLocal<T> target)
+                {
+                    return _weak.TryGetTarget(out target);
+                }
+
+                public WeakReferenceToLightThreadLocal(LightThreadLocal<T> instance)
+                {
+                    _hashCode = instance.GetHashCode();
+                    _weak = new WeakReference<LightThreadLocal<T>>(instance);
+                }
+
+                public bool Equals(WeakReferenceToLightThreadLocal other)
+                {
+                    if (ReferenceEquals(null, other)) return false;
+                    if (ReferenceEquals(this, other)) return true;
+                    if (_hashCode != other._hashCode)
+                        return false;
+                    if (_weak.TryGetTarget(out var x) == false ||
+                       other._weak.TryGetTarget(out var y) == false)
+                        return false;
+                    return ReferenceEquals(x, y);
+                }
+
+                public override bool Equals(object obj)
+                {
+                    if (ReferenceEquals(null, obj)) return false;
+                    if (ReferenceEquals(this, obj)) return true;
+                    if (obj.GetType() != this.GetType()) return false;
+                    return Equals((WeakReferenceToLightThreadLocal)obj);
+                }
+
+                public override int GetHashCode()
+                {
+                    return _hashCode;
+                }
+            }
+
+            public void Register(LightThreadLocal<T> parent)
+            {
+                _parents.Add(new WeakReferenceToLightThreadLocal(parent));
+            }
+
+            ~CurrentThreadState()
+            {
+                foreach (var parent in _parents)
+                {
+                    if (parent.TryGetTarget(out var liveParent) == false)
+                        continue;
+                    var copy = liveParent._values;
+                    if (copy == null)
+                        continue;
+                    if (copy.TryRemove(this, out var value)
+                        && value is IDisposable d)
+                    {
+                        d.Dispose();
+                    }
+                }
+            }
+        }
+        public void Dispose()
+        {

```

Comment:
We are missing `GC.SuppressFinalize(this);` here

Reply:
Good catch, yes

Reply:
I've fixed that